### PR TITLE
Avoid losing ops in file-based recovery

### PR DESCRIPTION
**Pull request by jasontedor**
When a primary is relocated from an old node to a new node, it can have ops in its translog that do not have a sequence number assigned. When a file-based recovery is started, this can lead to skipping these ops when replaying the translog due to a bug in the recovery logic. This commit addresses this bug and adds a test in the BWC tests.

Relates #22484

### CODE REVIEW

File: qa/backwards-5.0/src/test/java/org/elasticsearch/backwards/IndexingIT.java [hunk 1]
```diff
@@ -109,26 +112,58 @@ public void testSeqNoCheckpoints() throws Exception {
...
             assertSeqNoOnShards(nodes, checkGlobalCheckpoints, numDocsOnNewPrimary, newNodeClient);
         }
     }
 
+    private void assertCount(final String index, final String preference, final int expectedCount) throws IOException {
+        final Response response = client().performRequest("GET", index + "/_count", Collections.singletonMap("preference", preference));
+        assertOK(response);
+        final InputStream content = response.getEntity().getContent();
+        final int actualCount =
+            Integer.parseInt(XContentHelper.convertToMap(JsonXContent.jsonXContent, content, false).get("count").toString());
```
Comment (bleskes) [inline qa/backwards-5.0/src/test/java/org/elasticsearch/backwards/IndexingIT.java]: Any chance to use something like ObjectPath.evaluate(shard, "seq_no.local_checkpoint")?

File: qa/backwards-5.0/src/test/java/org/elasticsearch/backwards/IndexingIT.java [hunk 2]
```diff
@@ -109,26 +112,58 @@ public void testSeqNoCheckpoints() throws Exception {
...
-            int numDocs = indexDocs(index, 0, randomInt(5));
+            int numDocs = 0;
+            final int numberOfInitialDocs = 1 + randomInt(5);
+            logger.info("indexing [{}] docs initially", numberOfInitialDocs);
+            numDocs += indexDocs(index, 0, numberOfInitialDocs);
             assertSeqNoOnShards(nodes, checkGlobalCheckpoints, 0, newNodeClient);
-
             logger.info("allowing shards on all nodes");
             updateIndexSetting(index, Settings.builder().putNull("index.routing.allocation.include._name"));
             ensureGreen();
```
Comment (bleskes) [inline qa/backwards-5.0/src/test/java/org/elasticsearch/backwards/IndexingIT.java]: shall we add assert counts here too?

**Review by bleskes — CHANGES_REQUESTED** [review]
This LGTM. It would be great to have tests added in RecoverySourceHandlerTests .

**Comment by jasontedor** [comment]
@bleskes I pushed a unit tests in 4a010875bc988bc576bf0a9bb6cec4d9616576e5.

File: qa/backwards-5.0/src/test/java/org/elasticsearch/backwards/IndexingIT.java (hunk 1 above)
Comment (jasontedor) [inline qa/backwards-5.0/src/test/java/org/elasticsearch/backwards/IndexingIT.java]: I pushed 4430caa54a2b09f05f98331dc806e579d58aad52.

File: qa/backwards-5.0/src/test/java/org/elasticsearch/backwards/IndexingIT.java (hunk 2 above)
Comment (jasontedor) [inline qa/backwards-5.0/src/test/java/org/elasticsearch/backwards/IndexingIT.java]: I pushed 0112f909a6a427f78bb9a5fcaaae7fc829863151.

**Comment by jasontedor** [comment]
Thanks @bleskes. I've pushed commits in response to your comments.

**Review by bleskes — APPROVED** [review]
awesome

**Comment by bleskes** [comment]
test this please

**Comment by bleskes** [comment]
test this please

**Comment by jasontedor** [comment]
Thanks @bleskes.

**Pull request merged**